AGS 3: cherry picked and tidied mgambrell's AudioChannelsLocks commit

ivan-mogilko opened the pull request:
This pull request does following:

1. Cherry-picked @mgambrell's commit with channel locks from #659, with following edits:
* fixed few cases of broken logic, incorrect changes to array size etc;
* reduced big changes to indentation;
* reverted move of big chunks of code (not related to the commit's purpose);
the last two edits have a purpose to make merge cleaner.

2. Ammended this with little more refactoring, to comply with recent changes to audio code.

I believe that we need two more things here:
* Fix mutexes to work recursively on Linux too
* There are several places in soundclip implementations (OGG, MP3, MIDI) where AudioChannelsLock has to be placed, see TODO comments.
* anything else?

sonneveld commented:
The two new channel functions will need locks too.

ivan-mogilko commented:
> The two new channel functions will need locks too.

Which ones?

sonneveld commented:
> Which ones?

The Channel_is_playing and channel_has_clip ones. I thought this story was meant to prevent accidental use of `channels` without locks?

ivan-mogilko commented:
I double checked now, and I've put locks in them:
https://github.com/adventuregamestudio/ags/pull/689/commits/57ff5eab5462fd0259043cfbc4a884e660751059#diff-a80a847ddff0dbe95ff3bc30c9390f7dR70
Is this correct?

sonneveld commented:
Oh sorry, I thought I was looking at the file in github, I must have been looking at the original.

sonneveld commented on Engine/ac/global_audio.cpp:
```diff
@@ -354,24 +370,33 @@ void PlayMP3File (const char *filename) {
 
-    if ((channels[useChan] = my_load_static_ogg(asset_name, 150, doLoop)) != NULL) {
-        channels[useChan]->play();
+    SOUNDCLIP *clip = my_load_static_ogg(asset_name, 150, doLoop);
```
Nice. This was weird before as we'd overwrite the channel with null on failure.

ivan-mogilko replied:
It still overwrites channel with null on failure, in the last else below, just explicitly.

sonneveld commented on Engine/ac/global_audio.cpp:
```diff
@@ -386,21 +411,26 @@ void PlaySilentMIDI (int mnum) {
-    if (channels[play.silent_midi_channel] == NULL)
+
+    AudioChannelsLock lock;
+    auto* ch = lock.SetChannel(play.silent_midi_channel,load_sound_clip_from_old_style_number(true, mnum, false));
```
Should you do the same thing here? Not overwrite the channel on failure? I suppose we quit() on failure this time.

ivan-mogilko replied:
It seems like channels were usually overwritten with null on failure. I don't know if this is requirement so tried to keep same logic.

As for quitting, this is very inconsistent in AGS. I recieved requests to replace them with warnings to the log on numerous occasions. IMHO this should be configurable, in other words an "assert"-like function that either prints into log or quits depending on setup. Then we may send a message to the IDE if it's connected in debug session too, and even break at this point.

sonneveld commented on Engine/game/savegame.cpp:
```diff
@@ -555,6 +555,9 @@ HSaveError DoAfterRestore(const PreservedParams &pp, const RestoredData &r_data)
+    {
+    AudioChannelsLock lock;
     // NOTE: channels are array of MAX_SOUND_CHANNELS+1 size
     for (int i = 0; i <= MAX_SOUND_CHANNELS; ++i)
```
Potential for setting AudioChannelsLock within the lock.
edit: you might be doing a more global lock here because you want a consistent save state.

ivan-mogilko replied:
This is DoAfterRestore function where save data is already loaded and now applied. I am not sure if it's a good idea to make lock more global here because this function does many kinds of stuff, including *loading another room*.

sonneveld commented on Engine/main/update.cpp:
```diff
@@ -248,14 +248,16 @@ void update_overlay_timers()
-  const bool is_voice = channel_has_clip(SCHAN_SPEECH);
+  AudioChannelsLock lock;
+  auto *ch = lock.GetChannel(SCHAN_SPEECH);
+  const bool is_voice = ch != nullptr;
```
`is_voice` here could be the value of `is_playing`, then we won't need the lock for the rest of the function.

ivan-mogilko replied:
I am not sure, it does different actions depending on whether clip exists and whether it is playing below.
But I may read necessary values and unlock at once.

sonneveld commented on Engine/main/update.cpp:
```diff
@@ -291,7 +293,9 @@ void update_speech_and_messages()
-  const bool is_voice = channel_has_clip(SCHAN_SPEECH);
+  AudioChannelsLock lock;
+  auto *ch = lock.GetChannel(SCHAN_SPEECH);
+  const bool is_voice = ch != nullptr;
```
Similarly here, you could grab the speech pos here and free the lock.

ivan-mogilko replied:
Right, so I thought about that, and my next pr with nonblocking voice-over adds different way to test for voice mode, so some of these checks become unnecessary and the place when you need to access channel also changes, so I think I'll leave this as it is for now (some code is going to be removed anyway)

sonneveld commented on Engine/media/audio/ambientsound.cpp:
```diff
@@ -12,9 +12,7 @@
 //
 //=============================================================================
 
-#include "media/audio/ambientsound.h"
```
I tried to keep `audio_system.h` use out of `media/audio/*`

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -39,8 +39,53 @@
+//-----------------------
+//sound channel management; all access goes through here, which can't be done without a lock
+
+static std::array<SOUNDCLIP *,MAX_SOUND_CHANNELS+1> _channels;
```
I don't know if this gets null initialised or not for std::array.

ivan-mogilko replied:
If I understood correctly from the answer: https://stackoverflow.com/questions/18295302/default-initialization-of-stdarray

We need to add empty initializer list `{}` to initialize all members to default, which is nullptr supposedly in this case.

mgambrell replied:
that topic doesnt address your question, which is affected by the fact that statics are zero-initialized. in this case std::array is probably built with a standard array internally allocated not-on-the-heap, and the standard array is zero-initialized due to the whole thing being static. 

As for whether you ought to initialize static variables, that's a matter of elegance and what type of aesthetics you like and what kind of confidence you want to instill on the reader.  To me, leaving off the initializer means 'danger! must think about what it means that there was no initializer was here'. In this case, the reader is supposed to know that statics are zero-initialized, but I can see why it might be doubtful in this case.

A rule where arrays should always be initialized would hamper performance. And if there's any case where variables can be confidently not initialized, it's global statics.

ivan-mogilko replied:
Alright, got it.

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -39,8 +39,53 @@
+
+SOUNDCLIP *AudioChannelsLock::SetChannel(int index, SOUNDCLIP* ch)
+{
+    _channels[index] = ch;
```
Is it worth printing a debug warning here if we're overwriting a channel pointer? The channel would not get cleaned up otherwise. 

Another check might be to ensure we're not reusing a channel pointer, which I _think_ we're careful about but we'd have to check the cross fading code (that moves soundclips in/out of the music channel)

I'm fairly sure we're careful about freeing the channel first but it would be a good sanity check.

ivan-mogilko replied:
I can add warnings, but this situation is complicated, because clips could be saved elsewhere and not deleted.
I do not see any good solution here except for storing SOUNDCLIP in a smart pointer (either unique_ptr or shared_ptr, depending on what kind of manipulations we want to allow).

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -91,15 +136,18 @@ void start_fading_in_new_track_if_applicable(int fadeInChannel, ScriptAudioClip
-    channels[SPECIAL_CROSSFADE_CHANNEL] = channels[currentChannel];
-    channels[currentChannel] = NULL;
+
+    auto cfade_clip = lock.SetChannel(SPECIAL_CROSSFADE_CHANNEL, lock.GetChannel(currentChannel));
```
Maybe we need a MoveChannel() function to ensure no duplicate sound clips?

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -541,22 +602,26 @@ int get_volume_adjusted_for_distance(int volume, int sndX, int sndY, int sndMaxD
 void update_directional_sound_vol()
 {
-    for (int chan = 1; chan < MAX_SOUND_CHANNELS; chan++) 
+    AudioChannelsLock lock;
```
Potential for AudioChannelsLock within loop.

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -541,22 +602,26 @@ int get_volume_adjusted_for_distance(int volume, int sndX, int sndY, int sndMaxD
+{
+    AudioChannelsLock lock;
 
     for (int chan = 1; chan < MAX_SOUND_CHANNELS; chan++) {
```
Potential for AudioChannelsLock within loop.

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -592,9 +657,9 @@ void update_ambient_sound_vol () {
-        if (channel_is_playing(thisSound->channel)) {
-            channels[thisSound->channel]->set_volume(wantvol);
-        }
+        auto *ch = lock.GetChannel(thisSound->channel);
```
`GetChannelIsPlaying` here?

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -634,23 +699,26 @@ void shutdown_sound()
+    AudioChannelsLock lock;
+
     // find a free channel to play it on
     for (int i = SCHAN_NORMAL; i < MAX_SOUND_CHANNELS; i++) {
```
Potential for AudioChannelsLock within loop? We interact with another channel but it's later.

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -795,15 +867,21 @@ extern volatile char want_exit;
 
+//this is called at various points to give streaming logic a chance to update
+//it seems those calls have been littered around and points where it ameliorated skipping
+//a better solution would be to forcibly thread the streaming logic
```
Hah yes, we know.

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -918,41 +998,46 @@ void update_music_volume() {
                 if (crossFading > 0) {
-                    channels[SCHAN_MUSIC] = channels[crossFading];
-                    channels[crossFading] = nullptr;
+                    lock.SetChannel(SCHAN_MUSIC, lock.GetChannel(crossFading));
```
Another potential spot for a MoveChannel function.

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -963,8 +1048,8 @@ int prepare_for_new_music () {
                 stop_and_destroy_channel_ex(SCHAN_MUSIC, false);
-                channels[SCHAN_MUSIC] = channels[crossFading];
-                channels[crossFading] = NULL;
+                lock.SetChannel(SCHAN_MUSIC, lock.GetChannel(crossFading));
```
Another spot for a potential channel move function.

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -1049,36 +1134,33 @@ void play_new_music(int mnum, SOUNDCLIP *music) {
+    useChannel = prepare_for_new_music();
+    play.cur_music_number = mnum;
     current_music_type = 0;
-    channels[useChannel] = NULL;
```
Are we assuming prepare_for_new_music sets this to null? should we assert?

ivan-mogilko replied:
Yes, prepare_for_new_music has this in the end:

<pre>
// Just make sure, because it will be overwritten in a sec
    if (_lock.GetChannel(useChannel) != nullptr)
        stop_and_destroy_channel (useChannel);
</pre>

mgambrell commented:
Sorry, I don't really have time to evaluate this. But in general I tried not to think about things too much. And I can tell you, I probably put the locks outside the loops because I didn't want to analyze what happens if the loops get interrupted (probably nothing) but also because sound APIs should be atomic from the user's point of view. If someone says "fade out all the channels" he definitely doesn't mean "Fade out all the channels, but it's OK if some of them start fading a little bit after the others because random timing flukes". It's worth evaluating for loops that may take a long time, but not stuff that's effectively instantaneous anyway. Actually it's best if the entire frame's worth of audio commands are atomic too, but we're not gonna have that any time soon.

ivan-mogilko commented:
@mgambrell 
> I probably put the locks outside the loops because I didn't want to analyze what happens if the loops get interrupted (probably nothing) but also because sound APIs should be atomic from the user's point of view.

Yes, that makes sense. I am thinking to redo some of my changes related to loops.

sonneveld commented:
> And I can tell you, I probably put the locks outside the loops because I didn't want to analyze what happens if the loops get interrupted (probably nothing) but also because sound APIs should be atomic from the user's point of view.

Agree about the atomic sound api but as per our discussion a while ago, maybe that's worth considering when we can implement it that way.

I also just realised that it won't help polling that much, the polling loop will just block at the first channel that's locked and will wait before doing the rest anyway.

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -724,6 +791,8 @@ void clear_music_cache() {
 
 }
 
+void play_new_music(int mnum, SOUNDCLIP *music);
```
Defect: this is actually defined as static below and will fail to compile on some compilers.

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -634,23 +699,26 @@ void shutdown_sound()
         else if (!channel_is_playing(i)) {
             if (PlaySoundEx(val1, i) >= 0)
-                channels[i]->priority = priority;
+                ch->priority = priority;
```
defect: This code will segfault because `ch` is null but it is expecting channels[i] to have a value.  There may be other examples to watch out for.

ivan-mogilko replied:
And this is why in my old refactored branch I was returning locked *channels* instead of clips!

ivan-mogilko commented on Engine/ac/game.cpp:
```diff
@@ -1752,14 +1753,19 @@ void stop_fast_forwarding() {
+    AudioChannelsLock lock;
+
     // Restore actual volume of sounds
     for (int aa = 0; aa < MAX_SOUND_CHANNELS; aa++)
```
I have no idea, this code never changed since 2012. This needs separate investigation.

ivan-mogilko commented on Engine/ac/global_audio.cpp:
```diff
@@ -256,18 +256,28 @@ void scr_StopMusic() {
-            channels[crossFading]->seek (posn);
-        else if (channel_is_playing(SCHAN_MUSIC))
-            channels[SCHAN_MUSIC]->seek (posn);
+        AudioChannelsLock lock;
```
I think the logic remained the same:
* if crossFading > 0 then test crossFading channel first;
* if previous check did not succeed then test music channel.

I may simplify the code to reduce its size though.

sonneveld replied:
I do agree that the logic is the same, it's just not as obvious.  I tried something like this:

```
void SeekMP3PosMillis (int posn) {
    if (!current_music_type) { return; }

    AudioChannelsLock lock;

    auto *mus_ch = lock.GetChannel(SCHAN_MUSIC);
    auto *cf_ch = (crossFading > 0) ? lock.GetChannel(crossFading) : nullptr;

    if (cf_ch) {
        cf_ch->seek(posn);
    } else if (mus_ch) {
        mus_ch->seek(posn);
    }
}
```

sonneveld commented on Engine/ac/game.cpp:
```diff
@@ -269,18 +269,19 @@ void Game_SetAudioTypeVolume(int audioType, int volume, int changeType)
             {
-                channels[aa]->set_volume_percent(volume);
+                auto* ch = lock.GetChannel(aa);
+                ch->set_volume_percent(volume);
```
Needs to check if ch is null

ivan-mogilko replied:
It's relying on returned information about clip playing on this channel. All this is happening under one lock, so probably safe (at least I would think so).

ivan-mogilko replied:
It's relying on scrAudioChannel's information about clip playing on it. There might be more places like this, where engine does not consider that the state may be inconsistent.

mgambrell replied:
Oh yeah this is concerning. The existence of scrAudioChannel seems opposed to the safetiness of using my channel locking approach. It's like, more state, for each channel, stored separately. This really needs to be redesigned so it's kept together under control of the audio channel. could all the scrAudioChannel state be stored attached to the channel?

ivan-mogilko replied:
@mgambrell this is about something I did in my old attempt few years ago, where I tried to gather all channel-related data under one protected struct. I've mentioned it a bit ago in previous threads:
https://github.com/ivan-mogilko/ags-refactoring/commits/refactor_audio_channel

This was somewhat similar to your approach, except I was not returning clips, but *channels*, which was safer imho because clip may be replaced in the middle of function. For convenience IIRC I overloaded -> operator to access clip through this channel struct.

When these pull requests with threading audio was started to open I suggested to restore that branch, because it already contained lots of work, but then let myself get persuaded to not push too much refactoring now.
I keep getting very dissapointed by myself because of this.

ivan-mogilko replied:
The more issues is found in this pr, the more I think that maybe I should restore my old code instead, finish it and add mutexes, using your commit for reference.

sonneveld replied:
I mean, this PR is mostly ready. I can help add your channel refactoring after this one if you like.

sonneveld replied:
Another point is that unless the AudioChannel is an array of staticly allocated objects, you'll still need to lock the channels array when removing/adding channel pointers.

ivan-mogilko replied:
> Another point is that unless the AudioChannel is an array of staticly allocated objects, you'll still need to lock the channels array when removing/adding channel pointers.

AudioChannel actually is the array of statically allocated objects, they are created at engine startup and never changed.

ivan-mogilko replied:
> I mean, this PR is mostly ready. I can help add your channel refactoring after this one if you like.

Alright, question is: will additional null-pointer tests you suggested be enough to keep things at least thread safe?

sonneveld replied:
I believe so. You might be interested in this commit https://github.com/sonneveld/ags/commit/b2a6fe059cda5dd8be2526766a59f01a96aac8b7 I made some adjustments as I was reviewing this PR.

sonneveld commented on Engine/media/audio/audio.cpp:
```diff
@@ -918,41 +998,46 @@ void update_music_volume() {
             else {
                 if (crossFading > 0)
-                    channels[crossFading]->set_volume((curvol > targetVol) ? targetVol : curvol);
+                    lock.GetChannel(crossFading)->set_volume((curvol > targetVol) ? targetVol : curvol);
```
This grabs the channel without checking for null.   I only just noticed a couple of instance of these.. haven't thoroughly checked for others.

ivan-mogilko replied:
Here, and in other places, engine is relying on other variables like `crossFading`.

sonneveld replied:
Yes but the crossFading variable is not protected by the channel locks.

ivan-mogilko replied:
> Yes but the crossFading variable is not protected by the channel locks.

This is true...

sonneveld commented on Engine/ac/viewframe.cpp:
```diff
@@ -145,7 +145,11 @@ void CheckViewFrame (int view, int loop, int frame, int sound_volume) {
+    {
+        AudioChannelsLock lock;
+        auto* ch = lock.GetChannel(channel->id);
+        ch->set_volume_percent(ch->get_volume() * sound_volume / 100);
```
Worth checking if ch is null here since play_audio_clip_by_index can return null.

ivan-mogilko replied:
In that case `if (sound_volume != SCR_NO_VALUE && channel != NULL)` condition would be false

sonneveld replied:
I mean, that's technically correct, but it's a slightly round about way of checking it. Perhaps an assert check for the benefit of future readers?

ivan-mogilko commented:
I've mentioned this briefly in a code comment above, but want to state this more clearly.

Seeing all the issues found with this PR, for example related to potential inconsistency between different objects, such as ScriptAudioChannel and AudioChannel, I begin to think that I should have actually restored my old work from this branch and then ammend it with mutex, using new PRs for the reference:
https://github.com/ivan-mogilko/ags-refactoring/commits/refactor_audio_channel

I haven't check that code thoroughly for some time, but IIRC one of the advantages that it has is that it merged ScriptAudioChannel, AmbientSound and AudioChannel functionality in one object, reducing possible inconsistencies.

Another thing, it actually protected SOUNDCLIP* with smart pointer (we did not use C++11 back then so I wrote a quick emulation of unique_ptr).

Of course it is not necessarily matching the locks logic, but may be adjusted for it.

ivan-mogilko commented:
Or, we may keep it the way it is, except adding more null pointer tests as @sonneveld pointed out above, because we can no longer rely on additional variables and ScriptAudioChannels, which are not protected by locks....

ivan-mogilko commented:
Sorry for delay.

I followed the above suggestions, except the suggestion to move locks *inside* loops. On contrary I moved some of the locks I previously missed outside the loop to make whole loop an "atomic" operation (I think these loops usually do some fast processing).

* made sure that clip pointer is always tested after getting it from GetChannel();
* Added MoveChannel function which moves clip pointer from one channel to another (reduces code and maybe safer).
* followed suggestions from https://github.com/sonneveld/ags/commit/b2a6fe059cda5dd8be2526766a59f01a96aac8b7 (more or less)

sonneveld commented:
> I followed the above suggestions, except the suggestion to move locks _inside_ loops.

I diffed the old version of your branch with the new forced pushed one. All changes look good. It's interesting comparing our different style preferences :)

ivan-mogilko commented:
Is there anything I should add to this pr, like, fixed locks for Linux? And do we actually need locks in places like this:
https://github.com/adventuregamestudio/ags/pull/689/files#diff-a1c77e7c9d4359583ff8dde23b75e810R32

sonneveld commented:
> Is there anything I should add to this pr, like, fixed locks for Linux? And do we actually need locks in places like this: ..

I rebased my PR to be on top of this one. It includes the recursive locks and fixes the soundclip behaviour around destroy. https://github.com/adventuregamestudio/ags/pull/688
